cdap-api/src/main/java/io/cdap/cdap/api/messaging/Message.java

```diff
@@ -51,6 +56,23 @@ default String getPayloadAsString() {
     return getPayloadAsString(StandardCharsets.UTF_8);
   }
 
+  /**
+   * Decodes the message payload.
+   *
+   * @param decoder a {@link Function} to decode the object from a provided {@link Reader} that reads the payload
+   *                as {@link StandardCharsets#UTF_8} string.
+   * @param <T> the instance type of the decoded object
+   * @return the decoded object
+   */
+  default <T> T decodePayload(Function<Reader, T> decoder) {
+    try (Reader reader = new InputStreamReader(new ByteArrayInputStream(getPayload()), StandardCharsets.UTF_8)) {
+      return decoder.apply(reader);
+    } catch (IOException e) {
+      // This should never happen as we are reading from byte[]
```

Comment:
don't quite understand why this should never happen. Isn't it depending on what decoder is passed in ? 
I guess this assumption is particular to how this method is invoked currently ?